instrumentation/opentelemetry-instrumentation-sklearn/src/opentelemetry/instrumentation/sklearn/__init__.py

```diff
@@ -0,0 +1,738 @@
+            estimator (sklearn.base.BaseEstimator): A fitted ``sklearn``
+              estimator, typically a ``Pipeline`` instance.
+            attributes (dict): Attributes to attach to the spans.
+        """
+        if isinstance(estimator, self.exclude_classes):
+            logger.debug(
+                "Not instrumenting (excluded): %s",
+                estimator.__class__.__name__,
+            )
+            return
+
+        if isinstance(
+            estimator, tuple(self.recurse_namedtuple_attribs.keys())
+        ):
+            self._instrument_estimator_namedtuple(
+                estimator=estimator, attributes=attributes
+            )
+
+        if isinstance(estimator, tuple(self.recurse_attribs.keys())):
+            self._instrument_estimator_attribute(
+                estimator=estimator, attributes=attributes
+            )
+
+        for method_name in self.methods:
+            if hasattr(estimator, method_name):
+                self._instrument_instance_method(
+                    estimator=estimator,
+                    method_name=method_name,
+                    attributes=attributes,
+                )
+
+    def uninstrument_estimator(self, estimator: BaseEstimator):
+        """Uninstrument a fitted estimator and its hierarchy where configured.
+
+        Args:
+            estimator (sklearn.base.BaseEstimator): A fitted ``sklearn``
+              estimator, typically a ``Pipeline`` instance.
+        """
+        if isinstance(estimator, self.exclude_classes):
+            logger.debug(
+                "Not uninstrumenting (excluded): %s",
+                estimator.__class__.__name__,
+            )
+            return
+
+        if isinstance(
+            estimator, tuple(self.recurse_namedtuple_attribs.keys())
+        ):
+            self._uninstrument_estimator_namedtuple(estimator=estimator)
+
+        if isinstance(estimator, tuple(self.recurse_attribs.keys())):
+            self._uninstrument_estimator_attribute(estimator=estimator)
+
+        for method_name in self.methods:
+            if hasattr(estimator, method_name):
+                self._uninstrument_instance_method(
+                    estimator=estimator, method_name=method_name
+                )
+
+    def _check_instrumented(
+        self,
+        estimator: Union[BaseEstimator, Type[BaseEstimator]],
+        method_name: str,
+    ) -> bool:
+        """Check an estimator-method is instrumented.
+
+        Args:
+            estimator (BaseEstimator): A class or instance of an ``sklearn``
+              estimator.
+            method_name (str): The method name of the estimator on which to
+              check for instrumentation.
+        """
+        orig_method_name = "_otel_original_" + method_name
+        has_original = hasattr(estimator, orig_method_name)
+        orig_class, orig_method = getattr(
+            estimator, orig_method_name, (None, None)
+        )
+        same_class = orig_class == estimator
```

Comment:
Is this the check that prevents subclasses of an already instrumented parent class from instrumenting multiple times?

Reply:
Does this mean that the parent classes must be instrumented FIRST?

Reply:
> Is this the check that prevents subclasses of an already instrumented parent class from instrumenting multiple times?

This check prevents child classes from instrumenting an inherited method, which subsequently prevents multiple instrumentation. Whether or not the parent is already instrumented is irrelevant, so we don't have to worry about instrumenting in order or hierarchy.

> Does this mean that the parent classes must be instrumented FIRST?

No, it's written so that we only instrument non-inherited methods. Thus, order doesn't matter.